src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs

```diff
@@ -112,24 +114,60 @@ public override ValueTask<T> ReadAsync(CancellationToken cancellationToken)
 
             public override bool TryRead([MaybeNullWhen(false)] out T item)
             {
-                UnboundedChannel<T> parent = _parent;
+                UnboundedChannel<T, TQueue> parent = _parent;
+                return parent._items.IsThreadSafe ?
+                    IsThreadSafe(parent, out item) :
+                    NotIsThreadSafe(parent, out item);
 
-                // Dequeue an item if we can
-                if (parent._items.TryDequeue(out item))
+                static bool IsThreadSafe(UnboundedChannel<T, TQueue> parent, [MaybeNullWhen(false)] out T item)
                 {
-                    CompleteIfDone(parent);
-                    return true;
+                    if (parent._items.TryDequeue(out item))
+                    {
+                        CompleteIfDone(parent);
+                        return true;
+                    }
+
+                    item = default;
+                    return false;
                 }
 
-                item = default;
-                return false;
+                static bool NotIsThreadSafe(UnboundedChannel<T, TQueue> parent, [MaybeNullWhen(false)] out T item)
```

Comment:
again, another subjective rename proposal

```suggestion
                static bool NotThreadSafe(UnboundedChannel<T, TQueue> parent, [MaybeNullWhen(false)] out T item)
```